Automattic/woocommerce-payments
# Filter out non reusable gateways on the add payment method page

Closes https://github.com/Automattic/woocommerce-payments/issues/5022

#### Changes proposed in this Pull Request
> This PR is based on another one (https://github.com/Automattic/woocommerce-payments/pull/6901), which still needs to be merged as of Thu 3 Aug. 

On the _My account -> Payment methods -> Add payment method_ page, only reusable gateways should be listed. Nevertheless, for the split UPE & split UPE with deferred intent creation, this page also renders non-reusable gateways, leading to unpredictable behavior and multiple errors. The reason for that is that [the `is_add_payment_method_page()` function invoked here](https://github.com/Automattic/woocommerce-payments/blob/develop/includes/class-wc-payments.php#L767) returns the wrong result due to the plugin being not fully loaded at this point and the filtering doesn't occur as expected. `is_available` method within the gateway class is the right place to ensure the proper filtering.

If more details are needed, feel free to go through https://github.com/Automattic/woocommerce-payments/pull/6901's description, as the root cause of both issues is the same.


<details>
  <summary>Before & After</summary>

| Before  | After |
| ------------- | ------------- |
| <img width="594" alt="image" src="https://github.com/Automattic/woocommerce-payments/assets/22319444/185cdc84-ecec-4392-be5a-b4b1d0d83454">  |   <img width="594" alt="image" src="https://github.com/Automattic/woocommerce-payments/assets/22319444/013d2fc9-eab7-489c-8588-d2122967e667">  |


</details>










#### Testing instructions





0. Checkout `develop` & enable split UPE.
1. Enable multiple payment methods besides CC in _Payments -> Settings_.
2. Navigate to the _My account -> Payment methods -> Add payment method_ and confirm multiple loaded gateways and errors related to it.
3. Checkout `fix/filter-non-reusable-gateways-add-payment-method-page`, perform the previous step and confirm that only reusable gateways (e.g. CC) are displayed.
4. Enable split UPE with deferred intent creation, repeat step 2 and confirm that only reusable gateways are shown. 
5. Purchase a product using shortcode page and any gateway and confirm the order is successful (sanity check). 


-------------------


**Post merge**
